merkle/src/proof.rs

```diff
@@ -0,0 +1,150 @@
+pub struct DefaultMerkleProof<Item, Hash, Hasher, Merger>
+where
+    Hash: PartialEq + Clone,
+    Hasher: MerkleHasher<Item = Item, Hash = Hash>,
+    Merger: MerkleMerger<Hash = Hash>,
+{
+    index: usize,
+    item: Item,
+    peer_indices: Vec<usize>,
+    peer_hashes: Vec<Option<Hash>>,
+    _hasher: PhantomData<Hasher>,
+    _merger: PhantomData<Merger>,
+}
+
+impl<Item, Hash, Hasher, Merger> DefaultMerkleProof<Item, Hash, Hasher, Merger>
+where
+    Hash: PartialEq + Clone,
+    Hasher: MerkleHasher<Item = Item, Hash = Hash>,
+    Merger: MerkleMerger<Hash = Hash>,
+{
+    /// Creates merkle proof instance
+    /// index is the node index counted from root. e.g. index of the root node is 0
+    /// Note that index of a data node will always be positive
+    pub fn new(
+        index: usize,
+        item: Item,
+        peer_indices: Vec<usize>,
+        peer_hashes: Vec<Option<Hash>>,
+    ) -> Self {
+        assert!(index > 0, "index of a data node should always to positive");
+        assert_eq!(
+            peer_indices.len(),
+            peer_hashes.len(),
+            "length of peer_indices and peer_hashes should match"
+        );
+        assert!(
+            !peer_indices.is_empty(),
+            "length of peer_indices and peer_hashes should be positive"
+        );
+        Self {
+            index,
+            item,
+            peer_indices,
+            peer_hashes,
+            _hasher: Default::default(),
+            _merger: Default::default(),
+        }
+    }
+}
+
+impl<Item, Hash, Hasher, Merger> MerkleProof for DefaultMerkleProof<Item, Hash, Hasher, Merger>
+where
+    Hash: PartialEq + Clone + std::fmt::Debug,
+    Hasher: MerkleHasher<Item = Item, Hash = Hash>,
+    Merger: MerkleMerger<Hash = Hash>,
+{
+    type Hash = Hash;
+    type Error = MerkleProofError;
+
+    fn root_hash(&self) -> Result<Hash, Self::Error> {
+        // index of a data node should always to positive
+        if self.index > 0 {
+            // 1. Get the index and hash of the data node
+            let mut index = self.index;
+            let mut hash_opt = Some(Hasher::hash(
+                &self.item,
+                MerkleTreeNodeMetadata::new(self.index, self.peer_indices.len() as u32),
+            ));
+            for i in 0..self.peer_indices.len() {
+                // 2. Find the hash of its sibling
+                let peer_index = self.peer_indices[i];
+                let peer_hash_opt = &self.peer_hashes[i];
+                // 3 Prepare the input for for the hash merger by ordering the 2 hashes properly
+                let hashes = if index + 1 == peer_index {
+                    [hash_opt, peer_hash_opt.clone()]
+                } else if index == peer_index + 1 {
+                    [peer_hash_opt.clone(), hash_opt]
+                } else {
+                    return Err(MerkleProofError::InvalidProof);
+                };
```

Comment:
This would be great to refactor to its own function `get_ordered_siblings` or something

Reply:
Fixed